src/annoylib.h

```diff
@@ -178,11 +178,111 @@ struct Angular {
     // so we have to make sure it's a positive number.
     return sqrt(std::max(distance, T(0)));
   }
+  template<typename T>
+  static inline T pq_distance(T distance, T margin, int child_nr) {
+    if (child_nr == 0)
+      margin = -margin;
+    return std::min(distance, margin);
+  }
+  template<typename T>
+  static inline T pq_initial_value() {
+    return numeric_limits<T>::infinity();
+  }
+  static inline int chunkcount(int f) {
+    return f;
+  }
   static const char* name() {
     return "angular";
   }
 };
 
+struct Hamming {
+
+  template<typename S, typename T = int64_t>
+  struct ANNOY_NODE_ATTRIBUTE Node {
+    S n_descendants;
+    S children[2];
+    T v[1];
+  };
+
+  static inline size_t chunksize() {
+    return 64;
+  }
+  static const size_t max_iterations = 20;
+
+  template<typename T = int64_t>
+  static inline T pq_distance(T distance, T margin, int child_nr) {
+    return distance - (margin != child_nr);
+  }
+
+  template<typename T = int64_t>
+  static inline T pq_initial_value() {
+    return 0;
+  }
+  template<typename T = int64_t>
+  static inline T distance(const T* x, const T* y, int f) {
+    size_t dist = 0;
+    for (size_t i = 0; i < chunkcount(f); i++)
+    {
+      dist += __builtin_popcountll(x[i] ^ y[i]);
+    }
+    return dist;
+  }
+  template<typename S, typename T = int64_t>
+  static inline bool margin(const Node<S, T>* n, const T* y, int f) {
+    T chunk = n->v[0] / chunksize();
+    return (y[chunk] & (static_cast<T>(1) << (chunksize() - 1 - (n->v[0] % chunksize())))) != 0;
+  }
+  template<typename S, typename T = int64_t, typename Random>
+  static inline bool side(const Node<S, T>* n, const T* y, int f, Random& random) {
+    return margin(n, y, f);
+  }
+  static inline int chunkcount(int f) {
+    return f / chunksize() + 1;
```

Comment:
this should be `f / chunksize()` rounded up right? so i think it should be
`(f + chunksize() - 1) / chunksize()` instead?

Reply:
Yes, you are right! Fixing this.